File: code/backend/java/adapter-admin/src/main/java/games/paths/adapters/admin/catalog/FileCatalogWriter.java
package games.paths.adapters.admin.catalog;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import games.paths.adapters.rest.dto.StorySummaryResponse;
import games.paths.core.model.story.StorySummary;
import games.paths.core.port.story.CatalogWriterPort;

/**
 * FileCatalogWriter - v0.37.6 writes the static catalog under a local directory
 * (`game.catalog.export-dir`, env `CATALOG_EXPORT_DIR`); empty = not configured.
 */
@Component
public class FileCatalogWriter implements CatalogWriterPort {

    private final Path exportDir;
    private final ObjectMapper objectMapper;

    public FileCatalogWriter(@Value("${game.catalog.export-dir:}") String exportDir, ObjectMapper objectMapper) {
        this.exportDir = exportDir == null || exportDir.isBlank() ? null : Path.of(exportDir);
        this.objectMapper = objectMapper;
    }

    @Override
    public boolean isConfigured() {
        return exportDir != null;
    }

    @Override
    public String target() {
        return exportDir == null ? null : exportDir.toAbsolutePath().toString();
    }

    @Override
    public int write(String relativePath, List<StorySummary> stories) {
        // Same DTO and same ObjectMapper as GET /api/stories → byte-identical body.
        List<StorySummaryResponse> body = stories.stream().map(StorySummaryResponse::fromModel).toList();
        try {
            byte[] bytes = objectMapper.writeValueAsString(body).getBytes(StandardCharsets.UTF_8);
            Path file = exportDir.resolve(relativePath);
            Files.createDirectories(file.getParent());
            Files.write(file, bytes);
            return bytes.length;
        } catch (JsonProcessingException e) {
            throw new IllegalStateException("Cannot serialise the story catalog", e);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}

File: code/backend/python/app/adapters/catalog/file_catalog_writer.py
import json
from pathlib import Path
from typing import List
from fastapi.encoders import jsonable_encoder
from app.core.models.story.story_summary import StorySummary
from app.core.ports.story.catalog_writer_port import CatalogWriterPort


class FileCatalogWriter(CatalogWriterPort):
    """v0.37.6 — writes the static catalog under CATALOG_EXPORT_DIR; empty = not configured."""

    def __init__(self, export_dir: str | None):
        self.export_dir = Path(export_dir) if export_dir and export_dir.strip() else None

    def is_configured(self) -> bool:
        return self.export_dir is not None

    def target(self) -> str | None:
        return str(self.export_dir.resolve()) if self.export_dir else None

    def write(self, relative_path: str, stories: List[StorySummary]) -> int:
        # Same encoder + separators as FastAPI's JSONResponse → same body as GET /api/stories.
        body = json.dumps(jsonable_encoder(stories), ensure_ascii=False, separators=(",", ":"))
        data = body.encode("utf-8")
        file = self.export_dir / relative_path
        file.parent.mkdir(parents=True, exist_ok=True)
        file.write_bytes(data)
        return len(data)
